aten/src/ATen/core/jit_type.h

```diff
@@ -399,80 +434,83 @@ struct CAFFE2_API VaryingShape {
+      c10::optional<Device> device,
+      const VaryingShape<ShapeSymbol>& sizes,
+      const VaryingShape<int64_t>& strides,
+
+      // If we see `a + b + c`  and know that a, b, and c are the same size and have two dimensions (WxH), 
+      // then we can generate a fused kernel for them. 
+      // That fused kernel would likely have indexing math to handling both the W and H dimensions. 
+      // However, if we knew the WxH dimensions were contiguous, 
+      // we can pretend like we only have a single dimension, simplifying the indexing logic. 
+      // This can be performed even if the dimensions are transposed, 
+      // as long as a, b, and c are transposed in the same way. 
+      // We’d like to have the compiler be able to do this dimensionality reduction, 
+      // but simply knowing sizes is not enough. 
+      // We can extend profiling to also record stride information. 
+      // Rather than recording specific strides, 
+      // we can simply order the strides from smallest to largest with `stride_indices`
+      const VaryingShape<size_t>& stride_indices,
+      // A contiguity marker on the smallest stride (c0) indicates the stride is precisely 1, 
+      // otherwise a contiguity marker means that $stride_n = size_{n-1}*stride_{n-1}$
+      const VaryingShape<bool>& contiguity,
+      c10::optional<bool> requires_grad,
+      c10::optional<bool> undefined = false);
 
   static TensorTypePtr create(
       c10::optional<at::ScalarType> scalar_type,
       c10::optional<Device> device,
       c10::optional<size_t> dim,
-      c10::optional<bool> requires_grad) {
-    return TensorType::create(
-        scalar_type,
-        device,
-        VaryingShape(dim),
-        VaryingShape(dim),
-        requires_grad);
-  }
+      c10::optional<bool> requires_grad);
 
   // overloaded create variadic template argument as it could not distinguish
   // initializer list
   static TensorTypePtr createContiguous(
       at::ScalarType scalar_type,
       at::Device device,
-      at::IntArrayRef sizes) {
-    return create(
-        scalar_type,
-        device,
-        VaryingShape(sizes),
-        VaryingShape(contiguousStridesOf(sizes)),
-        c10::nullopt);
-  }
-  static TensorTypePtr create(
-      at::ScalarType scalar_type,
-      at::Device device,
-      at::IntArrayRef sizes,
-      at::IntArrayRef strides) {
-    return create(
-        scalar_type,
-        device,
-        VaryingShape(sizes),
-        c10::VaryingShape(strides),
-        c10::nullopt);
-  }
+      at::IntArrayRef sizes);
+
   static TypePtr fromNumberType(TypePtr typ);
   static TypePtr fromBoolType();
 
   c10::optional<size_t> dim() const {
     return sizes().size();
   }
 
-  const VaryingShape& sizes() const {
-    return sizes_;
-  }
-  const VaryingStrides& strides() const {
+  VaryingShape<int64_t> sizes() const; 
+
+  const VaryingShape<int64_t>& strides() const {
     return strides_;
   }
+
```

Comment:
The problem with representing stride_indices and continuity as separate VaryingShape elements is that it offers more ways in which the information is present/absent than what is actually expected. For instance, is there any case where we set some stride_index values but not the contiguity values? The current data structures can represent that, but it is not clear that was the intent vs. just a side effect of exposing the arrays as VaryingShape objects. In its current form any code that needs to use this information has to be much more verbose to handle a bunch of cases that don't actually happen. I still think it would be much better to create an abstract object that represents the striding information and documentation about what it means.